electrum/plugins/jade/jade.py

```diff
@@ -0,0 +1,464 @@
+    multisig_name = 'ele' + wallet_fingerprint_hash.hex()[:12]
+
+    # Collect all the signer data in case we need to register the
+    # multisig wallet on the Jade hw - NOTE: re-register is a no-op.
+    signers = []
+    for kstore in wallet.get_keystores():
+        fingerprint = kstore.get_root_fingerprint()
+        bip32_path_prefix = kstore.get_derivation_prefix()
+        derivation_path = bip32.convert_bip32_path_to_list_of_uint32(bip32_path_prefix)
+
+        # Jade only understands standard xtypes, so convert here
+        node = bip32.BIP32Node.from_xkey(kstore.xpub)
+        standard_xpub = node._replace(xtype='standard').to_xkey()
+
+        signers.append({'fingerprint': bytes.fromhex(fingerprint),
+                        'derivation': derivation_path,
+                        'xpub': standard_xpub,
+                        'path': []})
+
+    # Check multisig is registered - re-registering is a no-op
+    # NOTE: electrum multisigs appear to always be sorted-multisig
+    txin_type = wallet.get_txin_type(address)
+    keystore.register_multisig(multisig_name, txin_type, True, wallet.m, signers)
+
+    # Return the name used to register the wallet
+    return multisig_name
+
+class Jade_Client(HardwareClientBase):
+
+    @staticmethod
+    def _network() -> str:
+        return 'localtest' if constants.net.NET_NAME == 'regtest' else constants.net.NET_NAME
+
+    ADDRTYPES = {'standard': 'pkh(k)',
+                 'p2pkh': 'pkh(k)',
+                 'p2wpkh': 'wpkh(k)',
+                 'p2wpkh-p2sh': 'sh(wpkh(k))'}
+
+    MULTI_ADDRTYPES = {'standard': 'sh(multi(k))',
+                       'p2sh': 'sh(multi(k))',
+                       'p2wsh': 'wsh(multi(k))',
+                       'p2wsh-p2sh': 'sh(wsh(multi(k)))'}
+
+    @classmethod
+    def _convertAddrType(cls, addrType: str, multisig: bool) -> str:
+        return cls.MULTI_ADDRTYPES[addrType] if multisig else cls.ADDRTYPES[addrType]
+
+    def __init__(self, device: str, plugin: HW_PluginBase):
+        HardwareClientBase.__init__(self, plugin=plugin)
+
+        # Initial connection is with short timeout (to test availability)
+        self.jade = JadeAPI.create_serial(device, timeout=1)
+        self.jade.connect()
+
+        verinfo = self.jade.get_version_info()
+        self.fwversion = versiontuple(verinfo['JADE_VERSION'])
+        self.efusemac = verinfo['EFUSEMAC']
+
+        # Reconnecting with default timeout (for all subsequent calls)
+        self.jade = JadeAPI.create_serial(device)
+        self.jade.connect()
+
+        # Push some host entropy into jade
+        self.jade.add_entropy(os.urandom(32))
+
+    @runs_in_hwd_thread
+    def authenticate(self):
+        # Ensure Jade unlocked - always call hw unit at least once
+        # If the hw is already unlocked, this call returns immediately/no-op
+        authenticated = False
+        while not authenticated:
+            authenticated = self.jade.auth_user(self._network())
```

Comment:
It would be nice if we could show a message on screen to the user when they have to enter the PIN on the device, e.g. with `self.handler.show_message` as in the other user-interactions-prompts. I realise that `handler.show_message`/`handler.finished` does not compose with itself however, so this is not easy to do with how the code is currently written.
I guess it's fine as-is :)